compiler/syntax.go

```diff
@@ -379,6 +379,20 @@ func (c *syntaxLoader) convertSet(expr ast.SetExpression, nonterm *syntax.Nonter
 			default:
 				c.Errorf(op, "operator must be one of: first, last, precede or follow")
 			}
+		} else {
+			// Check if the reference targets another named set.
+			name := expr.Symbol().Name()
+			if index, ok := c.namedSets[name.Text()]; ok {
+				if index >= len(c.out.Sets) {
+					c.Errorf(expr.TmNode(), "forward references are not allowed")
```

Comment:
fwiw Forward references should work just fine - the circular dependencies are handled inside syntax/set.go.